crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -46,13 +85,269 @@ where
     HC: HashedCursor,
     VE: LeafValueEncoder<Value = HC::Value>,
 {
+    /// Takes a re-usable `RlpNode` buffer from the internal free-list, or allocates a new one if
+    /// the free-list is empty.
+    ///
+    /// The returned Vec will have a length of zero.
+    fn take_rlp_nodes_buf(&mut self) -> Vec<RlpNode> {
+        self.rlp_nodes_bufs
+            .pop()
+            .map(|mut buf| {
+                buf.clear();
+                buf
+            })
+            .unwrap_or_else(|| Vec::with_capacity(16))
+    }
+
+    /// Pushes a new branch onto the `branch_stack`, while also pushing the given leaf onto the
+    /// `child_stack`.
+    ///
+    /// This method expects that there already exists a child on the `child_stack`, and that that
+    /// child has a non-zero short key. The new branch is constructed based on the top child from
+    /// the `child_stack` and the given leaf.
+    fn push_new_branch(&mut self, leaf_key: Nibbles, leaf_val: VE::DeferredEncoder) {
+        // First determine the new leaf's shortkey relative to the current branch. If there is no
+        // current branch then the short key is the full key.
+        let leaf_short_key = if self.branch_stack.is_empty() {
+            leaf_key
+        } else {
+            // When there is a current branch then trim off its path as well as the nibble that it
+            // has set for this leaf.
+            leaf_key.slice_unchecked(self.branch_path.len() + 1, leaf_key.len())
```

Comment:
i see because the branch path length is < key length, so len() + 1 is in-bounds.